### ISSUE TITLE
User.activeAccount friction x session-keys

### ISSUE THREAD

**Issue by masvelio** [hicommonwealth/commonwealth#4809]
### Overview
This ticket describes the problems that I run into during the process of hooking up the new session-keys UI components to the current state. There are quite a lot of changes in the logic how the addresses should be handled and I guess it was not properly planned as from my research it requires refactoring of the application core. This is my hypothesis and the goal of this ticket should be to put everyone on the same page in terms of strategy for session-key implementation and any other upcoming project that touches login/user flow and controllers.

### Frictions
| Current behaviour  | Desired behaviour |
|--|--|
| If user is looking on non-community-related page like dashbaord or profile, the user.activeAccount/user.activeAccounts are not set | We should keep the state of active account all the time, regardless of the page being currently visited |
|If user is looking on non-community-related page like dashbaord or profile, the user dropdown list with addresses is not visible |We should display the user dropdown list with addresses regardless of the page being currently visited (this one is related to the previous row because the list is based on activeAccount/s)|
| If user is looking on community-related page, the user dropdown list with addresses displays only those addresses that are based on this chain (eg dydx - ethereum, chihuahua - cosmos) & that are joined to this community | The user dropdown list with addresses should display every unique address that is connected to the user profile, regardless of the chain. |
|If user connects a new address, this address is automatically joined to the community| The user should be able to add the address but the address won't be automatically joined to the community. User has to `join` it manually clicking the join button |

These are the things that I discovered so far. Probably I am not aware of many more cases that were not thought out.
The first question that came to my mind is what we can do to unblock session-keys project? But when I started looking at the code more deeply, I started asking question if this is a good time for session-keys project at all? Maybe we should first do some cleanup in the code and then start from the beginning => so revamp login flow and then (or simultaneously) take care of session keys? At first, I had the impression (and so was the task described) that session-keys is mainly blocked by the UI. In my opinion, however, session-keys is just one piece of the big puzzle and we cannot do this in separation.

### Possible solutions 

| Solution  |Pros|Cons|
|--|--|--|
|Adjust current implementation of user controller and make `activeAccount/s` work properly for the new requirements|Probably quickest way to unblock further work and deliver session-keys project|- Shortcut - A temporary fix for a problem that will come back to us in a moment. - Probably more things that `activeAccount/s` need rework which is not considered in this doc
|Start discussing about signin revamp (prioritise it)|- Very related to session-keys so we will be able to plan the work better - The opportunity for this kind of refactoring happens once every few years so it's worth considering the wider application area right away |- Development will stand still for a while, as we will have to focus on scheduling work - session keys project will not be delivered anytime soon
|Start moving user/login/session logic&controllers to react-query and zustand|- We need to do this anyway but definitely it would be better to do it before the big work like login revamp or session keys - alternatively, we can try to leave old controlers as they are and start implementing new login flow / session keys using react-query & new components, eg with feature flag |foundation work, users do not like it but it pays off for the company in long term

Whatever we choose (maybe some more ideas will come up) we need to really spend more time on technical planning as login flow and identity management seems to be most crucial thing in the app and it seems to be great moment to reiterate on it.

 
### Additional resources 
- [slack convo that started it all](https://commonxyz.slack.com/archives/C054J622P6J/p1692028654551569)
- #4314 main bucket ticket for session keys
- [figma file](https://www.figma.com/file/mJMPTzEnpEFIaR2wBWkElq/Session-Keys-and-Login?type=design&node-id=256%3A60411&mode=design&t=LVnGNoEn3HcQCrJj-1) for user dropdown list with addresses I was talking about above

**Comment by jnaviask** [comment]
cc @zakhap to review

**Comment by raykyri** [comment]
I don't believe these are significant enough concerns to justify a refactor. We are not at a point where we understand the new login requirements enough to start rearchitecting around them. 

If information is not available on pages without activeAccount, we should hide the relevant UI.

"Join community behavior" is entirely unrelated to session keys, as we don't require any session key interaction to join a community (yet).

**Comment by masvelio** [comment]
> We are not at a point where we understand the new login requirements enough to start rearchitecting around them.

Fair enough, @zakhap mentioned that this is something we will be working on soon so I wanted to make sure that we discuss if this is right time or not. 

> I don't believe these are significant enough concerns to justify a refactor.

For me it is enough, we are moving away from controllers into the react-query so we have to do the refactor anyways. I am not insisting it has to be right now, but I want us to take a broader perspective. Imo the code related to the login flow is a big mess and we have great opportunity to clean it up by refactoring it to the new approach. 

> If information is not available on pages without activeAccount, we should hide the relevant UI.

Not sure what case specifically you are talking about here but I am trying to figure out how we should modify the code to meet new design requirements - not other way round. Product/Design team wants to always show the list of addresses, not only on community-related pages. With current implementation of `user.activeAccounts` it is not possible, so my main question is if we want to modify current logic or write the new one. `user.activeAccount/s` have over 100 usages in the code so I am assuming that it will be hard to make it work properly across the board after modification. On the other hand writing new logic and have "two sources of truth" for active accounts does not seems right neither. 

Tldr, I can keep writing the code and we will see what happen (no idea when), but I would prefer to have an idea and at least rough plan how to tackle it to both keep the code untangled and meet new product requirements.

**Comment by masvelio** [comment]
I was also trying to figure out what would be the possibility to divide the process of `connecting a new address` into two steps (connect address and join community). Tried to track down where the joining community part happens but the login flow for me is a maze. I am also missing the domain knowledge how the login flow works and what are the relations between user/account/address/roles etc which makes me keep banging my head on the wall.

**Comment by masvelio** [comment]
My another idea to push session-keys project forward is that we will find some compromise with the Product/Design team and we cut the scope a bit. For example:
- we use current `user.activeAccount/s` implementation so the user dropdown address will be enhanced with the session-key status, but the addresses list will behave the same as now, so it will be only visible on community-related-pages and only with community-chain-related-addresses
- we keep the current logic for `connecting a new address` so the address will be automatically joined to the community

These two are most problematic things I have encountered so far. Not sure about other dependencies. Just thinking out loud.

**Comment by jnaviask** [comment]
I am super in favor of refactoring / revamping the way that activeAccount(s) works -- it's sorely needed for a long time. I agree with @raykyri that we should not block Session Key work delivery on that refactor, especially as we're still working out login/account-related requirements from the Product/Design side (cc @zakhap).

My stance is that for the time being we should proceed with this solution:

> Adjust current implementation of user controller and make activeAccount/s work properly for the new requirements

And see if we can make any quick wins along the way, while we begin planning for a larger refactor in parallel, with work items to be executed following the release of session key support.

@masvelio I think the two down-scope items (user dropdown has session-key status but address list doesn't change + keep logic for connecting a new address) are acceptable for the time being, although "don't automatically join to community" would be a very easy change to make on the backend (we already have the logic in place in the specific case where a community has a TOC -- then the logging-in user must click join and accept the TOC to properly join a community).

Can you document these changes to scope on the relevant tickets? And cc @zakhap again to ensure he is informed about these changes.

**Comment by jnaviask** [comment]
Would also like to flag @gdjohnson who may have an idea if the login flow is documented (also cc @mzparacha who last worked on it), to reduce the amount of research/digging that Marcin needs to do in order to move the ball forward here.

**Comment by masvelio** [comment]
@jnaviask my understanding is that if we decide to cut the scope (how the dropdown with addresses behaves) then the modification of `user.activeAccount/s` might not be needed . But maybe I am missing something 🤔 My idea was to cut the scope to prevent big code changes and push session-keys forward. Ray's PR is opened since April, so maybe we could just deliver MVP for session-keys so the foundation will be on master, and then we can build next iterations on top of that?

**Comment by gdjohnson** [comment]
@jnaviask No documentation in the wiki AFAICT. I do remember some Figma flowcharts mapping all the different login flow possibilities, though. 

Here's one that might be useful: https://www.figma.com/file/CXikWetKsXepvdtSFbJedB/Login-Flows?type=whiteboard&node-id=0%3A1&t=haxfv6y5qprIC7DO-1

**Comment by masvelio** [comment]
as agreed on the sync call, to move things forward we are gonna cut the scope so the user dropdown with addresses will use current implementation of `user.activeAccount/s`. This means that user dropdown will behave same as currently:
- will show addresses section only when user visits community-related page
- will show only addresses related to current community

Regarding second concern, per @jnaviask comment above, we should be able to split the logic of connecting new address + joining the community into to separate steps, so we will meet new requirements.
